File: node/sort/merge-sort.js
var numOfItems = process.argv[2];
var upperBound = Math.pow(2, 32);

var a = [];
for(var i  = 0; i < numOfItems; i++) {
    a.push(
        Math.floor(
            Math.random()*(upperBound + 1)
        )
    )
}

mergeSort(a);

function mergeSort(a) {
    if (a.length === 1) {
        return a;
    }

    var left = [];
    var right = [];
    for(var i = 0; i < a.length; i++) {
        if (i < parseInt(a.length / 2)) {
            left.push(a[i]);
        }
        else {
            right.push(a[i]);
        }
    }

    left = mergeSort(left);
    right = mergeSort(right);

    return merge(left, right)
}

function merge(left, right) {
    var result = [];
    var numOfIterations = left.length + right.length;
    var lastLeft = left.shift();
    var lastRight = right.shift();
    for(i = 0; i < numOfIterations; i++) {
        if (!!lastLeft && !!lastRight) {
            if (lastLeft >= lastRight) {
                result.push(lastRight);
                lastRight = right.shift();
                continue;
            }
            else {
                result.push(lastLeft);
                lastLeft = left.shift();
                continue;
            }
        }

        if (!!!lastLeft) {
            result.push(lastRight);
            lastRight = right.shift();
            continue;
        }

        if (!!!lastRight) {
            result.push(lastLeft);
            lastLeft = left.shift();
        }
    }

    return result;
}

File: python/sort/merge-sort.py
import random
import sys
import math
from collections import deque

numOfItems = int(sys.argv[1])
upperBound = math.pow(2, 32)

a = []
for i in range(0, numOfItems) :
    a.append(random.randint(0, upperBound))

def mergeSort(a) :
    if (len(a) == 1):
        return a

    left = []
    right = []
    for i in range(0, len(a)) :
        if i < len(a) / 2 :
            left.append(a[i])
        else :
            right.append(a[i])

    left = mergeSort(left)
    right = mergeSort(right)

    return merge(left, right)

def merge(left, right) :
    result = []
    numOfIterations = len(left) + len(right)
    left = deque(left)
    right = deque(right)
    lastLeft = left.popleft()
    lastRight = right.popleft()
    for i in range(0, numOfIterations) :
        if lastLeft and lastRight:
            if lastLeft >= lastRight:
                result.append(lastRight)
                lastRight = right.popleft() if len(right) > 0  else None
                continue
            else :
                result.append(lastLeft)
                lastLeft = left.popleft() if len(left) > 0 else None
                continue

        if not lastLeft :
            result.append(lastRight)
            lastRight = right.popleft() if len(right) > 0 else None
            continue

        if not lastRight :
            result.append(lastLeft)
            lastLeft = left.popleft() if len(left) > 0 else None

    return result

mergeSort(a)
